RAJA Perf baseline data generation and results summary (#49)
llnl/benchmarks: docs/13_rajaperf/rajaperf.rst

dependencies in Git submodules. Thus, dependency versions are pinned to each
version of the Suite. Building the Suite requires an installation of CMake for
configuring a build, a C++17 compliant compiler to build the code, and an MPI
library installation to link against.

> The Suite can be run in a myriad of ways via as command-line options and their

Reviewer: ```suggestion
The Suite can be run in a myriad of ways via command-line options and their
```

Author: fixed

   #. **basic/MULTI_REDUCE** multiple reductions in a kernel, where number of reductions is set at run time *(single loop, irregular atomic contention, RAJA::forall API)*
   #. **basic/REDUCE_STRUCT** multiple reductions in a kernel, where number of reductions (6) is known at compile time *(single loop, multiple reductions, RAJA::forall API)*
   #. **basic/INDEXLIST_3LOOP** construction of set of indices used in other kernel executions *(single loops, vendor scan implementations, RAJA::forall API)*
   #. **comm/HALO_PACKING_FUSED** packing and unpacking MPI message buffers for point-to-point distributed memory halo data exchange for mesh-based codes *(overhead of launching many small kernels, GPU variants use RAJA::Workgroup concepts to execute multiple kernels with one launch)* 

.. _rajaperf_fom-label:

Figure of Merit
---------------

The figure of merit (FOM) for each kernel is determined by the problem size at
which the kernel *saturates* resources on a *single* compute node. That is,
the problem size at which a computational throughput curve becomes flat, with

> zero derivative, and beyond which a running larger problem sizes does not yield

Reviewer: ```suggestion
zero derivative, and beyond which running larger problem sizes does not yield
```

Author: fixed

such as GFLOP/s, is plotted on the vertical axis as a function of problem size
on the horizontal axis. Ideally, such a curve will be monotonically increasing
and asymptote to a flat, horizontal line. Then, the saturation point is the
problem size at which the derivative of the throughput curve becomes zero.
In reality, throughput curves are often non-monotonic or do not have a 
strictly zero derivative for all points beyond some problem size. Therefore, we
apply a simple median based smoothing algorithm to the throughput curve data
and heuristically estimate the saturation point based on the smoothed
throughput curve. The details of our approach are documented in the
``process_data.py`` script in the `RAJAPerf-Benchmark GitHub project<https://github.com/llnl/RAJAPerf-Benchmark>`_
repository, which we use in :ref:`rajaperf_results-label`

Lastly, we emphasize that we want the kernels to be run in an execution
environment that aligns with how they would run if used in a real application.
Thus, the Suite should be run **using multiple MPI ranks** so that all
resources on a compute node are being exercised in a way that is
representative of how an application would run.

> All applications that use RAJA use it in an the *MPI + X* parallel application

Reviewer: ```suggestion
All applications that use RAJA use it in the *MPI + X* parallel application
```

Author: fixed

it. Executing a script with no arguments will print a message describing
required arguments to the screen. 

.. _rajaperf_build_mi300a-label:

MI300A architecture
--------------------

To configure and build the code to generate baseline data on a system with 
AMD MI300A processors (i.e., El Capitan (ATS-4) architecture) discussed in
:ref:`rajaperf_results-label`, we ran the following commands::

  $ pwd
  path/to/RAJAPerf
  $ ./scripts/lc-builds/toss4_cray-mpich_amdclang.sh 9.0.1 6.4.3 gfx942
  $ cd build_lc_toss4-cray-mpich-9.0.1-amdclang-6.4.3-gfx942
  $ make -j

Specifically, we configured and compiled the code for execution using version
9.0.1 of the Cray MPICH MPI library and the AMD clang compiler with ROCm
version 6.4.3 targeting GPU compute architecture gfx942.

.. _rajaperf_build_h100-label:

H100 architecture
--------------------

To configure and build the code to generate baseline data on a system with
NVIDIA H100 processors discussed in :ref:`rajaperf_results-label`, we ran the
following commands::

  $ pwd
  path/to/RAJAPerf
  $ ./scripts/lc-builds/toss4_mvapich2_nvcc_gcc.sh 2.3.7 12.9.1 90 10.3.1

>   $ build_lc_toss4-mvapich2-2.3.7-nvcc-12.9.1-90-gcc-10.3.1

Reviewer: ```suggestion
  $ cd build_lc_toss4-mvapich2-2.3.7-nvcc-12.9.1-90-gcc-10.3.1
```

Author: fixed

  $ make -j

Specifically, we configured and compiled the code for execution using version
2.3.7 of the MVAPICH2 MPI library, version 12.9.1 of the nvcc compiler targeting

> GPU compute architecture sm_90, and version 10.3.1 of the GNU compile for

Reviewer: ```suggestion
GPU compute architecture sm_90, and version 10.3.1 of the GNU compiler for
```

Author: fixed

compiling code for host execution

.. _rajaperf_run-label:

Running
=======

> After the RAJA Performance Suite code is built, the executable will located in

Reviewer: ```suggestion
After the RAJA Performance Suite code is built, the executable will be located in
```

Author: fixed

the ``bin`` subdirectory of the build space.

To get information about how to run the code, use the *help option*::

  $ pwd
  path/to/RAJAPerf
  $ cd my-build
  $ ls bin
  rajaperf.exe

>   $ .bin/rajaperf.exe --help (or -h)

Reviewer: ```suggestion
  $ ./bin/rajaperf.exe --help (or -h)
```

Author: fixed

Each kernel and variant run generates a checksum value based on kernel execution
output, such as an output data array computed by the kernel. The checksum
depends on the problem size run for the kernel; thus, each checksum is 
computed at run time. Validation criteria are defined in terms of the checksum
difference between each kernel variant and problem size run and a corresponding

> reference variant, which, by default, is the first variant of a kernel that is

Reviewer 2: Do we specifically want to say that the reference variant for validation is Base_Seq?

Author: The reference variant will be Base_CUDA or Base_HIP based on what you are running. The run scripts exclude all other variants.

This generates throughput curve files for ``Base_HIP`` and ``RAJA_HIP``
variants of each kernel and summarizes the FOM (described in
:ref:`rajaperf_fom-label`) in a CSV file. These files will be located in the
directory specified by via the ``--output-dir`` option above. We've included
a selected set of the files in this repo.

.. csv-table:: FOM results for Priority 1 kernels run on MI300A in CPX mode
   :file: ./baseline_data/MI300A/RPBenchmark_tier1-CPX/FOM/combined_fom.csv
   :align: center
   :widths: auto
   :header-rows: 1

AMD MI300A throughput plots
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following table contains throughput plots for each kernel run as described
above on the MI300A architecture in SPX mode and CPX mode. Each plot has multiple
curves where GFLOP/sec (compute rate) is plotted as a function of problem size
(allocated bytes). The left column shows SPX mode. The right column shows CPX
mode. The legend in each plot indicates the curves shown. Each plot includes:

>   * Throuput curves for Base and RAJA variant(s) of the kernel (solid line

Reviewer: ```suggestion
  * Throughput curves for Base and RAJA variant(s) of the kernel (solid line
```

Author: fixed

+-----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------+

NVIDIA H100 throughput results
-------------------------------

For the H100 architecture, we present throughput results, where we run with
4 MPI ranks on a node -- one for each H100 GPU. We run each ``Priority 1``
kernel over a sequence of problem sizes such that the saturation point is
evident on its associated throughput curve.

We choose the smallest problem to use ~50,000 bytes of allocated memory and
the largest problem to use ~150MB of allocated memory, which is about 3 times
the L2-cache size on the H100 GPU.  The L2-cache is 50 MB
(50 * 1024 * 1024 = 52428800 bytes).

Note that for two of the kernels ``FEMSWEEP`` and ``MASS3DEA``, we ran a
different problem size range because these kernels don't clearly saturate.
For them, we chose the smallest problem to use ~1.6MB of allocated memory
and the largest problem to use ~300MB memory, which is about 6 times the 

> L-2 cache size.

Reviewer: ```suggestion
L2 cache size.
```
I think this is more consistent

Author: fixed

After building the code as described in :ref:`rajaperf_build_h100-label`, we
run the ``Priority 1`` kernels as follows::

  $ pwd
  path/to/RAJAPerf
  $ cd build_lc_toss4-mvapich2-2.3.7-nvcc-12.9.1-90-gcc-10.3.1
  $ ./run_tier_h100.sh spx tier1

This generates a directory named ``RPBenchmark_H100_tier1``, which contains
the results files for each kernel run over its range of problem sizes.

Then, we process the data for reporting the results here in a concise form
by running a Python script we provide::

  $ pwd
  path/to/RAJAPerf
  $ python3 path/to/process_data.py --root-dir path/to/build_lc_toss4-mvapich2-2.3.7-nvcc-12.9.1-90-gcc-10.3.1/RPBenchmark_H100_tier1 --output-dir path/to/build_lc_toss4-mvapich2-2.3.7-nvcc-12.9.1-90-gcc-10.3.1/RPBenchmark_H100_tier1/Output

> This generates throughput curve files for ``Base_HIP`` and ``RAJA_HIP``

Reviewer: ```suggestion
This generates throughput curve files for ``Base_CUDA`` and ``RAJA_CUDA``
```

variants of each kernel and summarizes the FOM (described in
:ref:`rajaperf_fom-label`) in a CSV file. These files will be located in the
directory specified by via the ``--output-dir`` option above. We've included
a selected set of the files in this repo.

.. csv-table:: FOM results for Priority 1 kernels run on H100
   :file: ./baseline_data/H100/RPBenchmark_tier1/FOM/combined_fom.csv
   :align: center
   :widths: auto
   :header-rows: 1

NVIDIA H100 throughput plots
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following table contains throughput plots for each kernel run as described
above for the H100 architecture. Each plot has multiple curves where GFLOP/sec
(compute rate) is plotted as a function of problem size (allocated bytes). 
The legend in each plot indicates the curves shown. Each plot includes:

>   * Throuput curves for Base and RAJA variant(s) of the kernel (solid line

Reviewer: ```suggestion
  * Throughput curves for Base and RAJA variant(s) of the kernel (solid line
```

Author: fixed

|    :width: 100 %                                                                                    |
|    :align: center                                                                                   |
+-----------------------------------------------------------------------------------------------------+
|                                                                                                     |
| .. figure:: baseline_data/RPBenchmark_H100_tier1/figures/Apps_VOL3D_flops.png                       |
|    :width: 100 %                                                                                    |
|    :align: center                                                                                   |
+-----------------------------------------------------------------------------------------------------+

.. _rajaperf_memory-label:

Memory Usage
============

For the RAJA Performance Suite Benchmark, we run each kernel over a sequence
of problem sizes to generate a throughput curve and, based on that, estimate
a saturation point. The memory usage for each entry in the sequence is 

> roughly the same for each kernel. However, there is no significant meaning to

Reviewer: ```suggestion
roughly the same for each kernel. However, there is no significant conclustion to draw
```

> take away from this since the memory usage of kernels like those in the Suite

Reviewer: ```suggestion
from this since the memory usage of kernels like those in the Suite
```
